crates/signer/src/manager/dirk.rs

```diff
@@ -0,0 +1,917 @@
+    constants::COMMIT_BOOST_DOMAIN,
+    signature::compute_domain,
+    signer::{BlsPublicKey, BlsSignature, ProxyStore},
+    types::{Chain, ModuleId},
+};
+use rand::Rng;
+use tonic::transport::{Channel, ClientTlsConfig};
+use tracing::{info, trace, warn};
+use tree_hash::TreeHash;
+
+use crate::{
+    error::SignerModuleError::{self, DirkCommunicationError},
+    proto::v1::{
+        account_manager_client::AccountManagerClient, lister_client::ListerClient,
+        sign_request::Id as SignerId, signer_client::SignerClient, Account as DirkAccount,
+        DistributedAccount, GenerateRequest, ListAccountsRequest, ResponseState, SignRequest,
+        UnlockAccountRequest,
+    },
+};
+
+#[derive(Clone, Debug)]
+enum WalletType {
+    Simple,
+    Distributed,
+}
+
+#[derive(Clone, Debug)]
+struct HostInfo {
+    domain: String,
+    participant_id: u64,
+}
+
+#[derive(Clone, Debug)]
+struct Account {
+    wallet: String,
+    name: String,
+    public_key: Option<BlsPublicKey>,
+    hosts: Vec<HostInfo>,
+    wallet_type: WalletType,
+    signing_threshold: u32,
+    is_proxy: bool,
+}
+
+impl Account {
+    pub fn complete_name(&self) -> String {
+        format!("{}/{}", self.wallet, self.name)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct DirkManager {
+    chain: Chain,
+    channels: HashMap<String, Channel>, // domain -> channel
+    accounts: HashMap<String, Account>, // pubkey -> account
+    unlock: bool,
+    secrets_path: PathBuf,
+    proxy_store: Option<ProxyStore>,
+}
+
+impl DirkManager {
+    pub async fn new_from_config(chain: Chain, config: DirkConfig) -> eyre::Result<Self> {
+        let mut tls_configs = Vec::new();
+
+        // Create a TLS config for each host
+        for host in config.hosts.clone() {
+            let mut tls_config = ClientTlsConfig::new().identity(config.client_cert.clone());
+
+            if let Some(ca) = &config.cert_auth {
+                tls_config = tls_config.ca_certificate(ca.clone());
+            } else {
+                trace!(?host.domain, "CA certificate for domain not found");
+            }
+
+            tls_config = tls_config.domain_name(host.domain.unwrap_or_default());
+            tls_configs.push(tls_config);
+        }
+
+        // Create a channel for each host, attempt to connect,
+        // and save it into a hashmap Host->Channel
+        let mut channels = HashMap::new();
+        for (i, tls_config) in tls_configs.iter().enumerate() {
+            let config_host = config.hosts[i].clone();
+            let domain = config_host.domain.unwrap_or_default();
+            match Channel::from_shared(config_host.url.to_string())
+                .map_err(|_| eyre::eyre!("Invalid Dirk URL"))?
+                .tls_config(tls_config.clone())
+                .map_err(|_| eyre::eyre!("Invalid Dirk TLS config"))?
+                .connect()
+                .await
+            {
+                Ok(ch) => {
+                    channels.insert(domain, ch);
```

Comment:
`domain` can be `None` in multiple hosts, so we should use another key. Maybe URL, but it could be a better option

Reply:
Even a vector and access it by index. But not sure if this is possible.

Reply:
I'm checking this right now but it's a bit more complex that what I thought, you are right we cannot use the name as it can be empty, but when you ask for the participants of a Distributed Account you get a struct with a name, not a url, so I cannot match there.

I can try by index but not sure if the order is deterministic, could lead to problems and it's harder to debug.

Reply:
I think I made it work in 66821e703e72a1c1136e6d91b7c1432bd81cd461